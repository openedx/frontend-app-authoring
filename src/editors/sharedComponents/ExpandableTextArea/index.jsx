import React from 'react';
import PropTypes from 'prop-types';
import TinyMceWidget from '../TinyMceWidget';
import { prepareEditorRef } from '../TinyMceWidget/hooks';
import './index.scss';

const ExpandableTextArea = ({
  value,
  setContent,
  error,
  errorMessage,
  ...props
}) => {
  const { editorRef, refReady, setEditorRef } = prepareEditorRef();

  if (!refReady) { return null; }
  return (
    <>
      <div className="expandable-mce error">
        <TinyMceWidget
          editorContentHtml={value}
          editorRef={editorRef}
          editorType="expandable"
          setEditorRef={setEditorRef}
          updateContent={setContent}
          {...props}
        />
      </div>
      {error && (
        <div className="text-danger-500 x-small">
          {props.errorMessage}
        </div>
      )}
    </>
  );
};

ExpandableTextArea.defaultProps = {
  value: null,
  placeholder: null,
  error: false,
  errorMessage: null,
};

ExpandableTextArea.propTypes = {
  value: PropTypes.string,
  setContent: PropTypes.func.isRequired,
  placeholder: PropTypes.string,
  error: PropTypes.bool,
  errorMessage: PropTypes.string,
};

export default ExpandableTextArea;
