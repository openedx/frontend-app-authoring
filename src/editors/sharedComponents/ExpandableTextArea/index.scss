.expandable-mce {
  .error {
    outline: 2px solid #CA3A2F;
  }

  .mce-content-body {
    padding: 10px;

    p {
      margin: 0;
    }

    blockquote {
      margin: 16px 40px;
    }
  }

  *[contentEditable="false"] {
    outline: 1px solid #D7D3D1;
  }

  *[contentEditable="true"] {
    outline: 1px solid #707070;

    &:focus, &:active {
      outline: 2px solid #000000;
    }
  }
}
