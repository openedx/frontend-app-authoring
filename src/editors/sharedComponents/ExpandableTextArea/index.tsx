import React from 'react';
import TinyMceWidget from '../TinyMceWidget';
import { prepareEditorRef } from '../TinyMceWidget/hooks';
import './index.scss';

interface Props extends Partial<React.ComponentProps<typeof TinyMceWidget>> {
  setContent: (content: string) => void;
  value?: string | null;
  placeholder?: string;
  error?: boolean;
  errorMessage?: string;
}

export const ExpandableTextArea = ({
  value = null, // TODO: why not default to '' ?
  setContent,
  error = false,
  errorMessage = '',
  ...props
}: Props) => {
  const { editorRef, setEditorRef } = prepareEditorRef();

  return (
    <>
      <div className="expandable-mce error">
        {/* @ts-ignore TODO: Remove this 'ts-ignore' once TinyMceWidget is converted to TypeScript */}
        <TinyMceWidget
          {...props}
          editorContentHtml={value}
          editorRef={editorRef}
          editorType="expandable"
          setEditorRef={setEditorRef}
          updateContent={setContent}
        />
      </div>
      {error && (
        <div className="text-danger-500 x-small">
          {errorMessage}
        </div>
      )}
    </>
  );
};

export default ExpandableTextArea;
