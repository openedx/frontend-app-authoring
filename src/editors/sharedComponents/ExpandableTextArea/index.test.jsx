import React from 'react';
import { shallow } from '@edx/react-unit-test-utils';
import ExpandableTextArea from '.';

// Per https://github.com/tinymce/tinymce-react/issues/91 React unit testing in JSDOM is not supported by tinymce.
// Consequently, mock the Editor out.
jest.mock('@tinymce/tinymce-react', () => {
  const originalModule = jest.requireActual('@tinymce/tinymce-react');
  return {
    __esModule: true,
    ...originalModule,
    Editor: () => 'TiNYmCE EDitOR',
  };
});

jest.mock('../TinyMceWidget', () => 'TinyMceWidget');

// Mock the TinyMceWidget
jest.mock('../TinyMceWidget/hooks', () => ({
  prepareEditorRef: jest.fn(() => ({
    editorRef: { current: { value: 'something' } },
    refReady: true,
    setEditorRef: jest.fn().mockName('hooks.prepareEditorRef.setEditorRef'),
  })),
}));

describe('ExpandableTextArea', () => {
  const props = {
    value: 'text',
    setContent: jest.fn(),
    error: false,
    errorMessage: null,
  };
  describe('snapshots', () => {
    test('renders as expected with default behavior', () => {
      expect(shallow(<ExpandableTextArea {...props} />).snapshot).toMatchSnapshot();
    });
    test('renders error message', () => {
      const wrapper = shallow(<ExpandableTextArea {...props} error errorMessage="eRRormeSsaGE" />);
      expect(wrapper.snapshot).toMatchSnapshot();
    });
  });
});
