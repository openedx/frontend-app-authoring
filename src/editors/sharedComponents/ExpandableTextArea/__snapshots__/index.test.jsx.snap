// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandableTextArea snapshots renders as expected with default behavior 1`] = `
<Fragment>
  <div
    className="expandable-mce error"
  >
    <TinyMceWidget
      editorContentHtml="text"
      editorRef={
        {
          "current": {
            "value": "something",
          },
        }
      }
      editorType="expandable"
      placeholder={null}
      setEditorRef={[MockFunction hooks.prepareEditorRef.setEditorRef]}
      updateContent={[MockFunction]}
    />
  </div>
</Fragment>
`;

exports[`ExpandableTextArea snapshots renders error message 1`] = `
<Fragment>
  <div
    className="expandable-mce error"
  >
    <TinyMceWidget
      editorContentHtml="text"
      editorRef={
        {
          "current": {
            "value": "something",
          },
        }
      }
      editorType="expandable"
      placeholder={null}
      setEditorRef={[MockFunction hooks.prepareEditorRef.setEditorRef]}
      updateContent={[MockFunction]}
    />
  </div>
  <div
    className="text-danger-500 x-small"
  />
</Fragment>
`;
